sdl_canvas.go

```diff
@@ -38,11 +65,45 @@ func (sdlCanvas *SDLCanvas) Height() int {
 
 // AddDrawable adds a struct that implements the eff.Drawable interface
 func (sdlCanvas *SDLCanvas) AddDrawable(drawable Drawable) {
-	sdlCanvas.drawables = append(sdlCanvas.drawables, drawable)
+	sdlCanvas.drawables = append(sdlCanvas.drawables, &sdlDrawable{drawable: drawable})
+}
+
+//RemoveDrawable removes struct from canvas that implements eff.Drawable
+func (sdlCanvas *SDLCanvas) RemoveDrawable(drawable Drawable) {
+	index := -1
+	for i, d := range sdlCanvas.drawables {
+		if d.drawable == drawable {
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		return
+	}
+
+	sdlCanvas.drawables = append(sdlCanvas.drawables[:index], sdlCanvas.drawables[index+1:]...)
+}
+
+//AddKeyUpHandler adds key up event handler to the canvas
+func (sdlCanvas *SDLCanvas) AddKeyUpHandler(handler KeyHandler) {
+	sdlCanvas.keyUpHandlers = append(sdlCanvas.keyUpHandlers, handler)
+}
+
+//AddKeyDownHandler adds key down event handler to the canvas
+func (sdlCanvas *SDLCanvas) AddKeyDownHandler(handler KeyHandler) {
+	sdlCanvas.keyDownHandlers = append(sdlCanvas.keyDownHandlers, handler)
 }
 
 // Run creates an infinite loop that renders all drawables, init is only call once and draw and update are called once per frame
 func (sdlCanvas *SDLCanvas) Run() int {
+	if sdlCanvas.width == 0 {
+		sdlCanvas.width = defaultWidth
+	}
+
+	if sdlCanvas.height == 0 {
+		sdlCanvas.height = defaultHeight
+	}
+
 	var err error
 	sdl.CallQueue <- func() {
```

Comment:
I still have trouble understanding how there isn't a race condition present between adding the function to the `sdl.CallQueue` and checking the potential error generated by calling `CreateWindow`.  From my understanding, the functions communicated via the `sdl.CallQueue` are locked to the main OS thread, however the `if err != nil` below isn't necessarily on that thread, and might occur before the corresponding function in the call queue is executed.

I know you are aware of this, but thought it wise to echo some of what we discussed yesterday in case anyone else had information to provide.

Here is a quick read regarding the design patter being used for the `sdl.CallQueue`.
https://github.com/golang/go/wiki/LockOSThread